### PR TITLE
Story #1173 Mobile UI: Batch user managing

### PR DESCRIPTION
**Pull request by jrbirch99**
The code associated with users and batches has been ported from udes_11. Units tests have also been ported and created, where needs be. 

This story relies on methods that are in the process of being ported in other stories: story/1172, story/1174. To use some of these methods this branch has been rebased with story/1174 and is compared to that branch so only the changes made in story/1173 are present in the PR. Once story/1172 has been merged to master this branch will need to be rebased with master. Not all the unit tests will pass until story/1174 has the methods from story/1172.

### CODE REVIEW

File: addons/udes_stock_picking_batch/models/stock_picking.py [hunk 1]
```diff
@@ -148,3 +158,35 @@ def _priorities_has_ready_pickings(self, priorities, picking_type_id):
             ("state", "=", "assigned"),
         ]
         return self.search_count(domain) >= 1
+
+    def batch_to_user(self, user):
+        """Throws error if picking is batched to another user
+        or if user already has a batch
+        creates batch and adds picking to it otherwise
+        """
```
Comment (LoKKeR96) [inline addons/udes_stock_picking_batch/models/stock_picking.py:166]: Nit: can we rephrase it a bit? It's a little hard to follow. It should state clearly what the intentions are and not just what it does, then follow with the exceptions.

Also, I think the file is missing  a new line at the end? Might be wrong on this

File: addons/udes_stock_picking_batch/models/stock_picking_batch.py [hunk 2]
```diff
@@ -575,3 +614,293 @@ def drop_off_picked(self, continue_batch, move_line_ids, location_barcode, resul
...
+                raise ValidationError(
+                    _("Found %d batches for the user, please contact " "administrator.")
+                    % len(batches)
+                )
+
+            batch = batches
+
+        return batch
+
+    def _check_user_id(self, user_id):
```
Comment (adamrpatrick) [inline addons/udes_stock_picking_batch/models/stock_picking_batch.py:645]: Nit: This method might be better named as `_get_user_id` as it's not really checking much

File: addons/udes_stock_picking_batch/models/stock_picking.py [hunk 3]
```diff
@@ -1,4 +1,4 @@
-from odoo import models, fields, api
+from odoo import models, fields, api, _
 from .common import PRIORITY_GROUPS
 from odoo.exceptions import ValidationError
```
Comment (LoKKeR96) [inline addons/udes_stock_picking_batch/models/stock_picking.py:4]: I think the story asks to add a function called get_batch_for_pick? is that still required?

File: addons/udes_stock_picking_batch/models/stock_picking_batch.py [hunk 4]
```diff
@@ -391,15 +424,21 @@ def _populate_next_tasks(
...
             tasks.append(task)
             if limit and len(tasks) >= limit:
                 break
-            move_lines = move_lines.filtered(lambda ml: ml.id not in task["move_line_ids"])
+            move_lines = move_lines.filtered(
+                lambda ml: ml.id not in task["move_line_ids"]
+            )
         return tasks
 
-    def _populate_next_task(self, move_lines, task_grouping_criteria, priority_ml=False):
```
Comment (LoKKeR96) [inline addons/udes_stock_picking_batch/models/stock_picking_batch.py:402]: Nit: Lots of formatting but this line is 90 characters in length, the rule as far as I am aware is 100 and then it is formatted. Can you check your black formatting script? It should use 100 as the line maximum length.

File: addons/udes_stock_picking_batch/models/stock_picking_batch.py [hunk 5]
```diff
@@ -575,3 +614,293 @@ def drop_off_picked(self, continue_batch, move_line_ids, location_barcode, resul
...
+        if batches:
+            if len(batches) > 1:
+                raise ValidationError(
+                    _("Found %d batches for the user, please contact " "administrator.")
+                    % len(batches)
+                )
+
+            batch = batches
+
+        return batch
```
Comment (adamrpatrick) [inline addons/udes_stock_picking_batch/models/stock_picking_batch.py:643]: Nit: Some of this logic can be tidied up e.g. if the search returns no results it will be an empty recordset. Perhaps something like this:
```
batches = PickingBatch.search(BLAH)

if len(batches) > 1:
    raise ValidationError(BLAH)

return batches or None
```

File: addons/udes_stock_picking_batch/models/stock_picking_batch.py [hunk 6]
```diff
@@ -575,3 +614,293 @@ def drop_off_picked(self, continue_batch, move_line_ids, location_barcode, resul
...
+
+        if not user_id:
+            raise ValidationError(_("Cannot determine the user."))
+
+        return user_id
+
+    def get_user_batches(self, user_id=None):
+        """Get all batches for user"""
+        if user_id is None:
+            user_id = self.env.user.id
```
Comment (adamrpatrick) [inline addons/udes_stock_picking_batch/models/stock_picking_batch.py:657]: This snippet is the same as `_check_user_id`.

File: addons/udes_stock_picking_batch/models/stock_picking_batch.py [hunk 7]
```diff
@@ -575,3 +614,293 @@ def drop_off_picked(self, continue_batch, move_line_ids, location_barcode, resul
...
+
+        Raise a ValidationError in case it cannot perform a search
+        or if multiple batches are found for the specified user.
+        """
+        PickingBatch = self.env["stock.picking.batch"]
+
+        user_id = self._check_user_id(user_id)
+        batches = PickingBatch.search(
+            [("user_id", "=", user_id), ("state", "=", "in_progress")]
+        )
```
Comment (adamrpatrick) [inline addons/udes_stock_picking_batch/models/stock_picking_batch.py:631]: This snippet is the same as `get_user_batches`.

File: addons/udes_stock_picking_batch/models/stock_picking_batch.py [hunk 8]
```diff
@@ -575,3 +614,293 @@ def drop_off_picked(self, continue_batch, move_line_ids, location_barcode, resul
...
+        if user_id is None:
+            user_id = self.env.user.id
+
+        if not user_id:
+            raise ValidationError(_("Cannot determine the user."))
+
+        return user_id
+
+    def get_user_batches(self, user_id=None):
+        """Get all batches for user"""
```
Comment (LoKKeR96) [inline addons/udes_stock_picking_batch/models/stock_picking_batch.py:655]: That's misleading, it is returning only running batches. Can we improve docstring please?

File: addons/udes_stock_picking_batch/models/stock_picking_batch.py (hunk 2 above)
Comment (LoKKeR96) [inline addons/udes_stock_picking_batch/models/stock_picking_batch.py:645]: Small nit: not necessary but would be nice to add docstring explaining the intentions.

File: addons/udes_stock_picking_batch/models/stock_picking_batch.py [hunk 9]
```diff
@@ -575,3 +614,293 @@ def drop_off_picked(self, continue_batch, move_line_ids, location_barcode, resul
...
+        pickings_to_remove.with_context(lock_batch_state=True).write(
+            {"batch_id": False, "u_reserved_pallet": False}
+        )
+        pickings_to_add.with_context(lock_batch_state=True).write({"batch_id": self.id})
+        self._compute_state()
+
+        return self
+
+
+def get_next_name(obj, code):
```
Comment (LoKKeR96) [inline addons/udes_stock_picking_batch/models/stock_picking_batch.py:868]: Could it be worth having a chat with a TL to see if this could live somewhere else? It feels wrong to have it only here but not a blocker for the review.

File: addons/udes_stock_picking_batch/models/stock_picking.py [hunk 10]
```diff
@@ -26,14 +26,18 @@ class StockPicking(models.Model):
...
         for picking in self:
             if picking.move_line_ids:
-                categories = picking.move_line_ids.mapped("location_id.u_location_category_id")
-                picking.u_location_category_id = categories if len(categories) == 1 else False
+                categories = picking.move_line_ids.mapped(
+                    "location_id.u_location_category_id"
+                )
+                picking.u_location_category_id = (
+                    categories if len(categories) == 1 else False
+                )
```
[the remark this answers is not in the document]
Comment (jrbirch99) [reply addons/udes_stock_picking_batch/models/stock_picking.py:37]: yep will do.

File: addons/udes_stock_picking_batch/models/stock_picking_batch.py (hunk 2 above)
Comment (jrbirch99) [reply addons/udes_stock_picking_batch/models/stock_picking_batch.py:645]: Not sure I understand, as it seems to me that it checks that there is a user_id input and will return the environment user if no input is given or raise an exception if it set to false.

File: addons/udes_stock_picking_batch/models/stock_picking.py (hunk 3 above)
Comment (jrbirch99) [reply addons/udes_stock_picking_batch/models/stock_picking.py:4]: get_batch_for_pick lives in the workflows controller class in udes_11 in udes_workflows/controllers/common, and calls batch_to_user. I think when we port workflows this method will be ported doesn't seem right to port it as part of this story.

Comment (LoKKeR96) [reply addons/udes_stock_picking_batch/models/stock_picking.py:4]: No problem there as long as TL and PO are happy.

File: addons/udes_stock_picking_batch/models/stock_picking.py (hunk 1 above)
Comment (jrbirch99) [reply addons/udes_stock_picking_batch/models/stock_picking.py:166]: I don't see any complaint in github for a newline at the end of the file. I have also check in my editor, I think it is fine :)

File: addons/udes_stock_picking_batch/models/stock_picking_batch.py (hunk 9 above)
Comment (jrbirch99) [reply addons/udes_stock_picking_batch/models/stock_picking_batch.py:868]: I think as it stands in udes_11 it is only called in the picking_batch class, so I think it makes sense to keep it in that file for now. Maybe in the future it can be moved.

File: addons/udes_stock_picking_batch/models/stock_picking_batch.py (hunk 4 above)
Comment (jrbirch99) [reply addons/udes_stock_picking_batch/models/stock_picking_batch.py:402]: The black formatting script was checked and it is now set at 100 a line maximum length

File: addons/udes_stock_picking_batch/models/stock_picking_batch.py (hunk 9 above)
Comment (LoKKeR96) [reply addons/udes_stock_picking_batch/models/stock_picking_batch.py:860]: Yeah I totally agree that it's only used here and it makes sense to live here, sometimes Miquel suggested me to move things in other places and this could be one of those times. I was only suggesting to ask a TL. I won't block the review either way so as soon as you're done and no one comments on this I will resolve.

File: addons/udes_stock_picking_batch/models/stock_picking_batch.py (hunk 4 above)
Comment (LoKKeR96) [reply addons/udes_stock_picking_batch/models/stock_picking_batch.py:402]: Can you run it again? I tried running it on the file in this commit and the output is different than what's in the last commit so it needs to be reformatted. If it doesn't change then the setting for line length isn't working.

File: addons/udes_stock_picking_batch/models/stock_picking_batch.py (hunk 2 above)
Comment (adamrpatrick) [reply addons/udes_stock_picking_batch/models/stock_picking_batch.py:624]: The way I read it is that it's returning a user_id therefore it's getting a user_id rather than checking. In reality you could argue that it's doing a bit of both.

**Review by adamrpatrick — APPROVED** [review]
